src/axom/core/tests/core_openmp_map.hpp

```diff
@@ -0,0 +1,275 @@
+// Copyright (c) 2017-2021, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#include "axom/core/execution/execution_space.hpp"
+#include "axom/core/execution/for_all.hpp"
+#include "axom/core/Map.hpp"
+#include "gtest/gtest.h" /* for TEST and EXPECT_* macros */
+#include <string>
+
+namespace axom
+{
+namespace internal
+{
+template <typename Key, typename T, typename Hash, typename Policy>
+experimental::Map<Key, T, Hash, Policy> init(int N, int len)
+{
+  experimental::Map<Key, T, Hash, Policy> test(N, len);
+  EXPECT_EQ(N * len, test.max_size());
+  EXPECT_EQ(0, test.size());
+  EXPECT_EQ(true, test.empty());
+  EXPECT_EQ(len, test.bucket_size());
+  EXPECT_EQ(N, test.bucket_count());
+  return test;
+}
+
+template <typename Key, typename T, typename Hash, typename Policy>
+void test_storage(experimental::Map<Key, T, Hash, Policy> &test)
+{
+  experimental::Map<Key, T, Hash, Policy> *test2 = &test;
+  axom::for_all<axom::OMP_EXEC>(
+    0,
+    test.max_size(),
+    AXOM_LAMBDA(IndexType idx) {
+      Key key = idx;
+      T value = key * 27;
+      auto ret_test = test2->insert(key, value);
+      EXPECT_EQ(true, ret_test.second);
+    });
+  EXPECT_EQ(false, test.empty());
+  axom::for_all<axom::OMP_EXEC>(
+    0,
+    test.max_size(),
+    AXOM_LAMBDA(IndexType idx) {
+      Key key = idx;
+      EXPECT_EQ(key * 27, test2->find(key).value);
+    });
+  //This should fail, since we're at capacity.
+  auto ret = test.insert(test.max_size(), 900);
+  EXPECT_EQ(false, ret.second);
+}
+
+template <typename Key, typename T, typename Hash, typename Policy>
+void test_brackets(experimental::Map<Key, T, Hash, Policy> &test)
+{
+  experimental::Map<Key, T, Hash, Policy> *test2 = &test;
+  axom::for_all<axom::OMP_EXEC>(
+    0,
+    test.size(),
+    AXOM_LAMBDA(IndexType idx) {
+      Key key = idx;
+      EXPECT_EQ(key * 27, (*test2)[key]);
+    });
```

Comment:
I don't see any reason why the following wouldn't work, 
but consider also expressing this as:

```cpp
experimental::Map<Key, T, Hash, Policy>& test2 = test;
```
which would allow for expressions like:
```cpp
EXPECT_EQ(key * 27, test[key]);
```

Reply:
It might also be good to express this as the following to help make it more obvious what you're testing 
```cpp
  Key key = idx;
  Value actual_val = test2[key];  // or Value& ?
  Value exp_val = key*27;
  EXPECT_EQ(exp_val, actual_val);
```